source/agora/consensus/BlockRewardCalc.d

```diff
@@ -0,0 +1,651 @@
+        // no block reward payout spans adjecent periods
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(19), 2, 3,
+            [BlockRewardsTup(Amount(14), 6), BlockRewardsTup(Amount(21), 6)]);
+        assert(getAmountArr([0, 0, 0, 0, 0, 6, 0, 0, 8, 0, 0, 9, 0, 0, 12, 0, 0, 0, 0]) == actual_rewards);
+
+        // no extra reward for the the last block of the period
+        // block reward payout spans adjecent periods
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(19), 2, 3,
+            [BlockRewardsTup(Amount(15), 5), BlockRewardsTup(Amount(25), 5)]);
+        assert(getAmountArr([0, 0, 0, 0, 0, 9, 0, 0, 11, 0, 0, 15, 0, 0, 5, 0, 0, 0, 0]) == actual_rewards);
+
+        // extra reward for the the last block of the period
+        // block reward payout spans adjecent periods
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(19), 2, 3,
+            [BlockRewardsTup(Amount(16), 5), BlockRewardsTup(Amount(26), 5)]);
+        assert(getAmountArr([0, 0, 0, 0, 0, 9, 0, 0, 12, 0, 0, 15, 0, 0, 6, 0, 0, 0, 0]) == actual_rewards);
+
+        // extra reward for the the last block of the period
+        // block reward payout spans adjecent periods
+        // periods have different length
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(22), 2, 3,
+            [BlockRewardsTup(Amount(17), 4), BlockRewardsTup(Amount(7), 3), BlockRewardsTup(Amount(15), 6)]);
+        assert(getAmountArr([0, 0, 0, 0, 0, 12, 0, 0, 9, 0, 0, 7, 0, 0, 6, 0, 0, 5, 0, 0, 0, 0]) == actual_rewards);
+
+        // Checking whether we conform to the whitepaper's
+        // coin generation rate of 27 BOA coins per 5 second for the first year.
+        actual_rewards = getTestRewards!"TotalValidatorReward"(Height(0), Height(22), 2, 3,
+        ConsensusConfig().validator_block_rewards, 100, 5.seconds);
+        assert(getAmountArr([0, 0, 0, 0, 0, 810000000, 0, 0, 810000000, 0, 0, 810000000, 0, 0,
+                             810000000, 0, 0, 810000000, 0, 0, 810000000, 0]) == actual_rewards);
+    }
+
+    /***************************************************************************
+
+    Returns the total validator reward after applying the penalty for
+    missing signatures.
+
+    Params:
+        height = the height of the block that will contain the CoinBase transaction
+        comply_perc = (#actual signatures / #expected signatures) * 100 on all
+                      the blocks for which we want to calculate the block reward
+        indent = number of spaces before log messages
+
+    Returns:
+        validator whitepaper reward - penalty on validators due to missing signature
+
+    ***************************************************************************/
+
+    public Amount getReducedValidatorReward (in Height height, double comply_perc, uint indent = 0) @safe const nothrow
```

Comment:
Let me tell you a floating point joke:
A man has a computer problem. He decides to use floating point. Now he has 2.00000000001 problems.
Usually, we use an integer with a fixed precision for this kind of use case.

Reply:
ehhmm, I tried to explain it here:
https://github.com/bosagora/agora/pull/1768#issuecomment-799178954

any comment on that?